README.md
# poker-odds

[![Latest npm version](https://img.shields.io/npm/v/poker-odds.svg)](https://www.npmjs.com/package/poker-odds)
[![Node version required](https://img.shields.io/node/v/poker-odds.svg)](https://www.npmjs.com/package/poker-odds)
[![Build Status](https://img.shields.io/travis/CookPete/poker-odds/master.svg)](https://travis-ci.org/CookPete/poker-odds)
[![Test Coverage](https://img.shields.io/codecov/c/github/cookpete/poker-odds.svg)](https://codecov.io/gh/CookPete/poker-odds)
[![Donate](https://img.shields.io/badge/donate-PayPal-blue.svg)](https://paypal.me/ckpt)

A lightweight command line tool for calculating poker hand probabilities. No dependencies. No huge data files.

### Installation

```bash
# yarn
yarn global add poker-odds

# npm
npm install -g poker-odds
```

### Usage

```bash
poker-odds AcKh KdQs   # any number of hands supported
                       # use .. for random cards, .... for a random hand

# options
-b, --board Td7s8d     # community cards
-i, --iterations 1000  # number of preflop simulations to run, default: 100000
-e, --exhaustive       # run all preflop simulations
-p, --possibilities    # show individual hand possibilities
-n, --no-color         # disable color output
-v, --version          # show version
-h, --help             # show help
```

Use `--board` or `-b` to define community cards.

![--board example](https://user-images.githubusercontent.com/1926029/35532782-5de9aaa0-0533-11e8-8bf0-2864f2bf0a9c.png)

Use `--exhaustive` or `-e` to run all preflop simulations. Note that this will take some time.

![--exhaustive example](https://user-images.githubusercontent.com/1926029/35533275-eaea7690-0534-11e8-88c9-15c993916e89.png)

Use `--possibilities` or `-p` to show all possible hand outcomes. Hand possibilities are shown by default if only one hand is defined.

![--possibilities example](https://user-images.githubusercontent.com/1926029/35532961-e73f5d18-0533-11e8-86f0-22c17c2d2dda.png)

### API

The method used to calculate probabilities can be imported and used directly in a JS/node project:

```js
import { calculateEquity } from 'poker-odds'

const hands = [['As', 'Kh'], ['Kd', 'Qs']]
const board = ['Td', '7s', '8d']
const iterations = 100000 // optional
const exhaustive = false // optional

calculateEquity(hands, board, iterations, exhaustive)
```

`calculateEquity()` returns an array of hands with the results of the simulations:

```json
[
  {
    "hand": [
      "Ac",
      "Kh"
    ],
    "count": 990,
    "wins": 803,
    "ties": 15,
    "handChances": [
      { "name": "high card", "count": 376 },
      { "name": "one pair", "count": 479 },
      { "name": "two pair", "count": 78 },
      { "name": "three of a kind", "count": 13 },
      { "name": "straight", "count": 44 },
      { "name": "flush", "count": 0 },
      { "name": "full house", "count": 0 },
      { "name": "four of a kind", "count": 0 },
      { "name": "straight flush", "count": 0 },
      { "name": "royal flush", "count": 0 }
    ],
    "favourite": true
  },
  {
    "hand": [
      "Kd",
      "Qs"
    ],
    "count": 990,
    "wins": 172,
    "ties": 15,
    "handChances": [
      { "name": "high card", "count": 351 },
      { "name": "one pair", "count": 463 },
      { "name": "two pair", "count": 77 },
      { "name": "three of a kind", "count": 13 },
      { "name": "straight", "count": 41 },
      { "name": "flush", "count": 45 },
      { "name": "full house", "count": 0 },
      { "name": "four of a kind", "count": 0 },
      { "name": "straight flush", "count": 0 },
      { "name": "royal flush", "count": 0 }
    ],
    "favourite": false
  }
]
```


CHANGELOG.md
### Changelog
All notable changes to this project will be documented in this file.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v1.0.1](https://github.com/CookPete/poker-odds/compare/v1.0.0...v1.0.1)
> 30 January 2018
- Remove debug log [`254ea63`](https://github.com/CookPete/poker-odds/commit/254ea63c0e0eadd4c0c416872ba03c1c42668b79)

#### v1.0.0
> 30 January 2018
- Update node version requirements [`65e4cdc`](https://github.com/CookPete/poker-odds/commit/65e4cdc53585a79869f04fe80ca5227ce0413d55)
- Initial commit [`3b4b9c4`](https://github.com/CookPete/poker-odds/commit/3b4b9c42d62c4bc5ccb186c6ea3c55849e3e1537)



test/rank.js
import test from 'ava'
import { rankHand, rankValues, getFlush } from '../src/rank'

test('high card', t => {
  t.is(rankHand(['As', '8d', '3d', '6s', '2c', 'Qd', '5h']), '0ec865')
  t.is(rankValues(['A', 'Q', '8', '6', '5', '3', '2']), '0ec865')
})

test('one pair', t => {
  t.is(rankHand(['As', 'Ad', '3d', '6s', '2c', 'Qd', '5h']), '1eec65')
  t.is(rankValues(['A', 'A', 'Q', '6', '5', '3', '2']), '1eec65')
})

test('two pair', t => {
  t.is(rankHand(['As', 'Ad', '3d', '3s', '2c', 'Qd', '5h']), '2ee33c')
  t.is(rankValues(['A', 'A', 'Q', '5', '3', '3', '2']), '2ee33c')
})

test('three of a kind', t => {
  t.is(rankHand(['As', 'Ad', 'Ad', '3s', '2c', 'Qd', '5h']), '3eeec5')
  t.is(rankValues(['A', 'A', 'A', 'Q', '5', '3', '2']), '3eeec5')
})

test('straight', t => {
  t.is(rankHand(['As', 'Qd', 'Kd', 'Js', 'Tc', 'Qd', '5h']), '4edcba')
  t.is(rankHand(['As', '4d', 'Ad', '3s', '2c', 'Qd', '5h']), '45432e')
  t.is(rankValues(['A', 'K', 'Q', 'Q', 'J', 'T', '5']), '4edcba')
  t.is(rankValues(['A', 'A', 'Q', '5', '4', '3', '2']), '45432e')
})

test('flush', t => {
  t.is(rankHand(['Ad', 'Qd', 'Kd', 'Js', 'Td', 'Qc', '5d']), '5edca5')
})

test('full house', t => {
  t.is(rankHand(['As', 'Ad', '3d', '6s', 'Ac', 'Qd', '3h']), '6eee33')
  t.is(rankValues(['A', 'A', 'A', 'Q', '6', '3', '3']), '6eee33')
})

test('four of a kind', t => {
  t.is(rankHand(['As', 'Ad', 'Ah', '6s', 'Ac', 'Qd', '3h']), '7eeeec')
  t.is(rankValues(['A', 'A', 'A', 'A', 'Q', '6', '3']), '7eeeec')
})

test('straight flush', t => {
  t.is(rankHand(['As', '3d', '4d', '6d', 'Ac', '5d', '7d']), '876543')
})

test('royal flush', t => {
  t.is(rankHand(['As', 'Ts', '4d', 'Js', 'Ks', '5d', 'Qs']), '9edcba')
})

test('invalid hand', t => {
  t.throws(() => rankHand(['As', 'Ad', 'Ah', 'Ac', 'Ac', 'Qd', '3h']))
  t.is(rankValues(['A', 'A', 'A', 'A', 'A', '6', '3']), undefined)
})

test('getFlush', t => {
  t.is(getFlush('ccddhhs'), undefined)
  t.is(getFlush('cddddds'), 'd')
  t.is(getFlush('ccccccc'), 'c')
})


test/utils.js
import test from 'ava'
import {
  numericalValue,
  numericalSort,
  convertToHex,
  parseCards,
  percent,
  seconds,
  getStraight,
  padStart,
  padEnd
} from '../src/utils'

test('numericalValue', t => {
  t.is(numericalValue('As'), 14)
  t.is(numericalValue('Td'), 10)
  t.is(numericalValue('5c'), 5)
  t.is(numericalValue('2h'), 2)
})

test('numericalSort', t => {
  const input = ['2h', 'Td', '5c', 'As', '5d']
  const expected = ['As', 'Td', '5c', '5d', '2h']
  t.deepEqual(input.sort(numericalSort), expected)
})

test('convertToHex', t => {
  t.is(convertToHex(['2h', 'Td', '5c', 'As', '5d']), '2a5e5')
})

test('parseCards', t => {
  t.deepEqual(parseCards('2hTd'), ['2h', 'Td'])
  t.deepEqual(parseCards('AdAc'), ['Ad', 'Ac'])
  t.deepEqual(parseCards('Ks..'), ['Ks', '..'])
  t.deepEqual(parseCards('....'), ['..', '..'])
  t.deepEqual(parseCards(''), undefined)
  t.deepEqual(parseCards('random string'), undefined)
})

test('percent', t => {
  t.is(percent(0.5), '50.0%')
  t.is(percent(0.12), '12.0%')
  t.is(percent(0.123), '12.3%')
  t.is(percent(0.1234), '12.3%')
  t.is(percent(0.1235), '12.4%')
  t.is(percent(0.001), '0.1%')
  t.is(percent(0.0000001), '0.1%')
  t.is(percent(0), '·')
})

test('seconds', t => {
  t.is(seconds(1), '1ms')
  t.is(seconds(500), '500ms')
  t.is(seconds(999), '999ms')
  t.is(seconds(1000), '1.0s')
  t.is(seconds(1100), '1.1s')
  t.is(seconds(1150), '1.2s')
  t.is(seconds(12345), '12.3s')
})

test('getStraight', t => {
  t.is(getStraight(['A', 'K', 'Q', 'J', 'T', '9', '8']), 'edcba')
  t.is(getStraight(['A', 'J', 'T', '9', '8', '7', '3']), 'ba987')
  t.is(getStraight(['A', '9', '7', '5', '4', '3', '2']), '5432e')
  t.is(getStraight(['A', 'K', 'Q', 'J', '9', '7', '3']), null)
  t.is(getStraight(['9', '8', '7', '5', '4', '3', '2']), null)
})

test('padStart', t => {
  t.is(padStart('hello', 10, 'a'), 'aaaaahello')
  t.is(padStart('hello', 10), '     hello')
  t.is(padStart('hello', 6), ' hello')
  t.is(padStart('hello', 5), 'hello')
  t.is(padStart('hello', 3), 'hello')
})

test('padEnd', t => {
  t.is(padEnd('hello', 10, 'o'), 'helloooooo')
  t.is(padEnd('hello', 10), 'hello     ')
  t.is(padEnd('hello', 6), 'hello ')
  t.is(padEnd('hello', 5), 'hello')
  t.is(padEnd('hello', 3), 'hello')
})


test/console.js
import test from 'ava'
import {
  color,
  colorCards,
  getHands,
  hasOption,
  getOption
} from '../src/console'

test('color', t => {
  t.is(color('hello', 'green'), '\x1b[32mhello\x1b[0m')
  t.is(color('hello', 'grey'), '\x1b[90mhello\x1b[0m')
  t.is(color('hello', 'beige'), 'hello')
  t.is(color('hello', null), 'hello')
})

test('colorCards', t => {
  t.is(colorCards(['As', 'Ad']), '\x1b[34mAs\x1b[0m \x1b[31mAd\x1b[0m')
  t.is(colorCards(['..', '..']), '\x1b[33m..\x1b[0m \x1b[33m..\x1b[0m')
  t.is(colorCards(['random', 'string']), 'random string')
})

test('getHands', t => {
  const argv = ['node', 'script.js', 'AsAd', 'KdQd', '--iterations', '100']
  t.deepEqual(getHands(argv), ['AsAd', 'KdQd'])
})

test('hasOption', t => {
  const argv = ['node', 'script.js', 'AsAd', 'KdQd', '--iterations', '100']
  t.is(hasOption('--iterations', argv), true)
})

test('getOption returns integer', t => {
  const argv = ['node', 'script.js', 'AsAd', 'KdQd', '--iterations', '100']
  t.is(getOption('--iterations', argv), 100)
})

test('getOption returns string', t => {
  const argv = ['node', 'script.js', 'AsAd', 'KdQd', '--board', '4s7d8c']
  t.is(getOption('--board', argv), '4s7d8c')
})

test('getOption returns undefined', t => {
  const argv = ['node', 'script.js', 'AsAd', 'KdQd', '--iterations', '100']
  t.is(getOption('--board', argv), undefined)
})

test('getOption supports shorthand option', t => {
  const argv = ['node', 'script.js', 'AsAd', 'KdQd', '-i', '100']
  t.is(getOption('--iterations', argv), 100)
})


src/rank.js
import lookup from '../lookup.json'
import { numericalSort, convertToHex, getStraight } from './utils'

export function rankValues (values) {
  let total = 0
  let max = 0
  const cardMatches = {}
  for (let i = 0; i !== values.length; i++) {
    cardMatches[values[i]] = 0
    for (let j = 0; j !== values.length; j++) {
      if (i === j) continue // TODO: Could this be i <= j?
      const first = values[i]
      const second = values[j]
      if (first === second) {
        cardMatches[first]++
        total++
        max = Math.max(cardMatches[first], max)
      }
    }
  }
  const matches = total / 2
  const straight = getStraight(dedupe(values)) // Dedupe to match straights like AKKKQJT
  const kickers = convertToHex(values.sort((a, b) => cardMatches[b] - cardMatches[a]))

  if (max > 3) {
    return undefined
  }
  if (max === 3) {
    return '7' + kickers.slice(0, 4) + getHighestKicker(kickers.slice(4)) // four of a kind
  }
  if (max === 2 && matches > 3) {
    return '6' + kickers.slice(0, 5) // full house
  }
  if (straight) {
    return '4' + straight // straight
  }
  if (max === 2) {
    return '3' + kickers.slice(0, 5) // three of a kind
  }
  if (max === 1 && matches > 1) {
    return '2' + kickers.slice(0, 4) + getHighestKicker(kickers.slice(4)) // two pair
  }
  if (max === 1) {
    return '1' + kickers.slice(0, 5) // one pair
  }
  return '0' + kickers.slice(0, 5) // high card
}

export function rankHand (input) {
  const hand = input.slice(0).sort(numericalSort)
  const values = hand.map(c => c[0]).join('')
  const suits = hand.map(c => c[1]).sort().join('')

  const rank = lookup.rank[values]
  const flush = lookup.flush[suits]

  if (!rank) {
    throw Error(`Invalid hand: ${hand.join(' ')}`)
  }

  const straight = rank[0] === '4'

  if (straight && flush) {
    const flushed = hand.filter(c => c[1] === flush).map(c => c[0])
    const kickers = getStraight(flushed)
    if (kickers) {
      // royal or straight flush
      return (kickers[0] === 'e' ? '9' : '8') + kickers
    }
  }
  if (flush) {
    // Fix kickers for flush
    // ie the highest cards of the flush suit
    const kickers = convertToHex(hand.filter(c => c[1] === flush).slice(0, 5))
    return '5' + kickers // flush
  }
  return rank
}

export function getFlush (string) {
  const match = string.match(/(s{5}|c{5}|d{5}|h{5})/)
  return match ? match[0][0] : undefined
}

function getHighestKicker (string) {
  return string.split('').sort().reverse()[0]
}

function dedupe (array) {
  return array.filter(function (item, index, array) {
    return array.indexOf(item) === index
  })
}


src/console.js
const HAND_PATTERN = /^[AKQJT2-9.][schd.][AKQJT2-9.][schd.]$/
const CONSOLE_COLORS = {
  black: '30',
  red: '31',
  green: '32',
  yellow: '33',
  blue: '34',
  magenta: '35',
  cyan: '36',
  white: '37',
  grey: '90'
}

export function color (string, color) {
  if (!color || !CONSOLE_COLORS[color] || hasOption('--no-color')) {
    return string
  }
  return `\x1b[${CONSOLE_COLORS[color]}m${string}\x1b[0m`
}

export function colorCards (cards) {
  return cards.map(colorCard).join(' ')
}

function colorCard (card) {
  if (/^.[sc]$/.test(card)) {
    return color(card, 'blue')
  }
  if (/^.[dh]$/.test(card)) {
    return color(card, 'red')
  }
  if (card === '..') {
    return color(card, 'yellow')
  }
  return card
}

export function getHands (argv = process.argv) {
  return argv.filter(string => HAND_PATTERN.test(string))
}

export function hasOption (option, argv = process.argv) {
  return argv.includes(option) || argv.includes('-' + option[2])
}

export function getOption (option, argv = process.argv) {
  const index = argv.indexOf(option)
  if (index === -1) {
    if (option.length !== 2) {
      return getOption('-' + option[2], argv)
    }
    return undefined
  }
  const value = argv[index + 1]
  if (/^\d+$/.test(value)) {
    return parseInt(value)
  }
  return value
}


src/calculate.js
import { RANK_NAMES } from './utils'
import { rankHand } from './rank'
import { createDeck, deal } from './deck'

export function calculateEquity (hands, board = [], iterations = 100000, exhaustive = false) {
  let results = hands.map(hand => ({
    hand,
    count: 0,
    wins: 0,
    ties: 0,
    handChances: RANK_NAMES.map(name => ({ name, count: 0 }))
  }))
  if (board.length === 5) {
    results = analyse(results, board)
  } else if (board.length >= 3) {
    const deck = createDeck(board.concat(...hands))
    for (let i = 0; i !== deck.length; i++) {
      if (board.length === 4) {
        results = analyse(results, board.concat(deck[i]))
        continue
      }
      for (let j = 0; j !== deck.length; j++) {
        if (i >= j) continue
        results = analyse(results, board.concat([ deck[i], deck[j] ]))
      }
    }
  } else if (exhaustive) {
    const deck = createDeck(board.concat(...hands))
    for (let a = 0; a !== deck.length; a++) {
      for (let b = 0; b !== deck.length; b++) {
        if (a <= b) continue
        for (let c = 0; c !== deck.length; c++) {
          if (b <= c) continue
          for (let d = 0; d !== deck.length; d++) {
            if (c <= d) continue
            for (let e = 0; e !== deck.length; e++) {
              if (d <= e) continue
              results = analyse(results, [deck[a], deck[b], deck[c], deck[d], deck[e]])
            }
          }
        }
      }
    }
  } else {
    for (let i = 0; i !== iterations; i++) {
      const randomCards = deal([].concat(...hands), 5 - board.length)
      results = analyse(results, board.concat(randomCards))
    }
  }
  const maxWins = Math.max(...results.map(hand => hand.wins))
  return results.map(hand => ({
    ...hand,
    favourite: hand.wins === maxWins
  }))
}

function analyse (results, board) {
  const ranks = results.map(result => {
    if (result.hand.includes('..')) {
      const randomCards = deal(board.concat(...results.map(r => r.hand)), 4)
      const hand = result.hand.map((card, index) => {
        if (card === '..') {
          return randomCards[index]
        }
        return card
      })
      return rankHand(hand.concat(board))
    }
    return rankHand(result.hand.concat(board))
  })
  const bestRank = ranks.slice(0).sort().reverse()[0]
  const tie = ranks.filter(rank => rank === bestRank).length > 1
  for (let i = 0; i !== results.length; i++) {
    if (ranks[i] === bestRank) {
      if (tie) {
        results[i].ties++
      } else {
        results[i].wins++
      }
    }
    results[i].count++
    results[i].handChances[parseInt(ranks[i][0])].count++
  }
  return results
}


src/utils.js
export const CARD_VALUES = ['2', '3', '4', '5', '6', '7', '8', '9', 'T', 'J', 'Q', 'K', 'A']
export const CARD_SUITS = ['s', 'c', 'h', 'd']
export const RANK_NAMES = [
  'high card',
  'one pair',
  'two pair',
  'three of a kind',
  'straight',
  'flush',
  'full house',
  'four of a kind',
  'straight flush',
  'royal flush'
]
const NUMERICAL_VALUES = {
  T: 10,
  J: 11,
  Q: 12,
  K: 13,
  A: 14
}
const STRAIGHTS = [
  'AKQJT',
  'KQJT9',
  'QJT98',
  'JT987',
  'T9876',
  '98765',
  '87654',
  '76543',
  '65432',
  '5432A'
]

export function numericalValue (card) {
  return NUMERICAL_VALUES[card[0]] || parseInt(card[0])
}

export function numericalSort (a, b) {
  return numericalValue(b) - numericalValue(a)
}

export function convertToHex (input) {
  input = typeof input === 'string' ? input.split('') : input
  return input
    .map(c => numericalValue(c).toString(16))
    .join('')
}

export function parseCards (string) {
  if (!string) {
    return undefined
  }
  return string.match(/[AKQJT2-9.][schd.]/g) || undefined
}

export function percent (number) {
  if (number === 0) {
    return '·'
  }
  if (number > 0 && number < 0.001) {
    return '0.1%'
  }
  return `${round(number * 100)}%`
}

export function seconds (ms) {
  if (ms >= 1000) {
    return `${round(ms / 1000)}s`
  }
  return `${ms}ms`
}

export function getStraight (hand) {
  const values = hand.join('')
  const suffix = values[0] === 'A' ? 'A' : '' // Append A to capture 5432A
  for (let i = 0; i !== STRAIGHTS.length; i++) {
    if (`${values}${suffix}`.includes(STRAIGHTS[i])) {
      return convertToHex(STRAIGHTS[i])
    }
  }
  return null
}

export function padStart (string, length, padString = ' ') {
  if (string.length >= length) {
    return string
  }
  return padString.repeat(length - string.length) + string
}

export function padEnd (string, length, padString = ' ') {
  if (string.length >= length) {
    return string
  }
  return string + padString.repeat(length - string.length)
}

function round (number, dp = 1) {
  const multiplier = dp * 10
  return (Math.round(number * multiplier) / multiplier).toFixed(dp)
}
